docs/adr/0006-consensus-governance.md

```diff
@@ -0,0 +1,478 @@
+
+Upon processing any proposal the following steps are then performed:
+
+- The `min_proposal_deposit` base units are transferred from the signer's
+  account to the governance service's _proposal deposit account_.
+
+- The signer's account is saved.
+
+- A new proposal is created and assigned an identifier.
+
+- The corresponding `ProposalSubmittedEvent` is emitted with the following
+  structure:
+
+  ```golang
+  type ProposalSubmittedEvent struct {
+      // ID is the unique identifier of a proposal.
+      ID uint64 `json:"id"`
+      // Submitter is the staking account address of the submitter.
+      Submitter staking.Address `json:"submitter"`
+  }
+  ```
+
+- The corresponding `staking.TransferEvent` is emitted, indicating transfer from
+  the submitter's account to the _proposal deposit account_.
+
+#### Vote
+
+Voting for submitted consensus layer governance proposals.
+
+**Method name:**
+
+```
+gov.CastVote
+```
+
+**Body:**
+
+```golang
+type CastVote struct {
+    // ID is the unique identifier of a proposal.
+    ID uint64 `json:"id"`
+    // Vote is the vote.
+    Vote Vote `json:"vote"`
+}
+```
+
+Upon processing a vote the following steps are performed:
+
+- The entity descriptor corresponding to the transaction signer is fetched. In
+  case no such entity exists, the method call fails with `ErrNotEligible`.
+
+- It is checked whether any entity's nodes are in the current validator set. In
+  case they are not, the method call fails with `ErrNotEligible`.
+
+- The proposal identified by `id` is loaded. If the proposal does not exist,
+  the method call fails with `ErrNoSuchProposal`.
+
+- If the proposal's state is not `StateActive`, the method call fails with
+  `ErrVotingIsClosed`.
+
+- The vote is added to the list of votes. If the vote already exists, it is
+  overwritten.
+
+- The corresponding `VoteEvent` is emitted with the following structure:
+
+  ```golang
+  type VoteEvent struct {
+      // ID is the unique identifier of a proposal.
+      ID uint64 `json:"id"`
+      // Submitter is the staking account address of the submitter.
+      Submitter staking.Address `json:"submitter"`
+      // Vote is the cast vote.
+      Vote Vote `json:"vote"`
+  }
+  ```
+
+### Queries
+
+This proposal introduces the following query methods in the governance module:
+
+```golang
+type Backend interface {
+    // ActiveProposals returns a list of all proposals that have not yet closed.
+    ActiveProposals(ctx context.Context, height int64) ([]*Proposal, error)
+
+    // Proposal looks up a specific proposal.
+    Proposal(ctx context.Context, query *ProposalQuery) (*Proposal, error)
+
+    // Votes looks up votes for a specific proposal.
+    Votes(ctx context.Context, query *ProposalQuery) ([]*VoteEntry, error)
+
+    // PendingUpgrades returns a list of all pending upgrades.
+    PendingUpgrades(ctx context.Context, height int64) ([]*upgrade.Descriptor, error)
```

Comment:
This should probably return some new type that also contains the information at which epoch the upgrade is scheduled for?

Reply:
The upgrade descriptor has that information already?

Reply:
Ah yeah, that's true :-)